src/oauth2/service.ts

```diff
@@ -142,7 +142,7 @@ export async function generateTokenForClient(client: OAuth2Client): Promise<OAut
  *
  * The resource owner then exchanges that code for an access and refresh token.
  */
-export async function generateTokenFromCode(client: OAuth2Client, code: string): Promise<OAuth2Token> {
+export async function generateTokenFromCode(client: OAuth2Client, code: string, codeVerifier: string|undefined): Promise<OAuth2Token> {
```

Comment:
I think it might be nice to make `codeVerifier` get a more specific type like `'S256' | 'plain'`, and make sure that this is validated as well. (earlier, in the controller)

Reply:
Here, `codeVerifier` refers to the verification code the client sends in the body of the token request _after_ the authorization request happens (where the challenge method is provided). This is only ever a string (if provided) or undefined (if not provided).

But this does bring up the fact that I'm not validating the `codeChallengeMethod` in the authorization request. I did think of that later. I forgot to add _any_ validation to the authorization request (such as providing the `codeChallenge` but not the `codeChallengeMethod`). I'll add a check there to verify that the provided `codeChallengeMethod` (if included) is either `plain` or `S256`.

Reply:
Cool!